### DISCUSSION TITLE
Nx integrated repo + Vitest workspace

### DISCUSSION THREAD

**Discussion by connorjs** [vitest-dev/vitest#4544] [Q&A] [answered]
Discussion thread for [Nx integrated repo](https://nx.dev/concepts/integrated-vs-package-based) and how the community currently uses Vitest with them.

I wanted a discussion, not an issue, to discuss ideas prior to formal "issues." See #4181 and #4519 for maybe-related issues.

**Comment by connorjs** [comment]
## Thread 1: Distributed vitest execution + single coverage aggregate

### Background

The [Nx mental model docs](https://nx.dev/concepts/mental-model) probably provide the best background, but for those not familiar with Nx (or similar tools like Turborepo), here’s my summary.

- Split your workspace/repository into projects (Vitest workspace == Nx workspace; Vitest project == Nx project)

- These projects DO NOT have to represent actual npm packages; they can exist simply for domain modeling or other organization
  - (In fact there does not even have to be a `package.json` in the integrated repo approach)

- The workspace defines a set of targets to run in each project (for this example, let’s say we have `build`, `lint`, and `coverage`)

- Nx provides mechanisms to run the targets across all projects: `nx run-many --targets build,lint,coverage`.

  It will also cache and only re-run work when needed. So, let’s say we have `libs/is-even`, `libs/is-odd`, and `apps/foo`. If I only change code in `apps/foo`, then running `nx run-many --targets build,lint,coverage` will NOT rerun any targets in either `libs` projects.

  _Nx extends this idea to the `affected` concept, but I will ignore that for now._

### Goal

So, our goal w.r.t. Vitest is to run `nx run-many -t coverage` many times a day and let Nx distribute/drive the execution in only the projects that need it.

### Solutions I see

#### Don’t use Nx here

Vitest has [workspace support](https://vitest.dev/guide/workspace.html), so just use it and ignore Nx.

> For aggregated results, Vitest expects a single vitest executable at the root repository/workspace level
>
> -- Paraphrased from [comment on 4181](https://github.com/vitest-dev/vitest/issues/4181#issuecomment-1817962022)

The consequence being that we’ve now lost out on Nx benefits.

**But, maybe Vitest caching delivers a similar local DX?**

- [The cache docs](https://vitest.dev/config/#cache) suggest that it does not do what we want here.
 
  > At the moment Vitest stores cache for test results to run the longer and failed tests first.

- [The changed command](https://vitest.dev/guide/cli.html#changed) could be used, but requires an actual commit to base from. This works, but not as well as Nx. _I will try this though._

- [The related command](https://vitest.dev/guide/cli.html#vitest-related) could be used, but I’m unsure how to "easily" do this. The perk of Nx is just running the command and trusting Nx to do the right thing.

**NOTE**: There's trade-offs at play. Nx won't re-run untouched projects, but will re-run an entire project that's touched. Vitest related and changed, hypothetically (I just haven't tested) could be more targeted/efficient.

#### Project argument + aggregation

The solution I tried to build was to run `vitest run` once for each project (`N` executions) and then have the coverage data either automatically merge or have a mechanism to merge/aggregate at the end.

[nyc (Istanbul)](https://github.com/istanbuljs/nyc#combining-reports-from-multiple-runs) suggests that it has this, so I may switch from V8 to try it out.

However, I imagine a world where the `reportersDirectory` is defined at the workspace level (I think it already is?) and then each project writes to a sub-path of it. I'm unsure how the `html` reporter works to comment on the feasibility, but I'm hopeful given `nyc` can do it and think the reporters are shared.

**Reply by connorjs** [reply]
#### Both Nx and vitest

TL;DR - Set up Nx to run the individual `coverage` executions. This works fine _minus_ aggregated coverage.

Let CI run a single vitest workspace run, which _will_ emit aggregated coverage.

This just requires clarifying how coverage thresholds work at workspace level and/or/vs project level.

**Reply by connorjs** [reply]
Some code snippets.

First, a `vitest.workspace.ts` file. I ended up not using this.

```ts
import { defineWorkspace } from "vitest/config";

const projects = [`apps/vite-spa`, `libs/hello-js`];

// eslint-disable-next-line import/no-default-export -- Vitest config
export default defineWorkspace(
	projects.map((project) => ({
		extends: `./vitest.config.ts`,
		test: { name: project, root: project },
	})),
);

// Alternative way to get projects, but requires cached project graph from "@nx/devkit"
// const projects = Object.values(readCachedProjectGraph().nodes)
// 	.filter((projectNode) =>
// 		projectNode.data.targets
// 			? Object.keys(projectNode.data.targets).includes(`test`)
// 			: false,
// 	)
// 	.map((projectNode) => projectNode.data.root);
```

Second, my actual planned `vitest.config.ts` (for now at least / for my current project)

```ts
import tsconfigPaths from "vite-tsconfig-paths";
import {
	configDefaults,
	coverageConfigDefaults,
	defineConfig,
} from "vitest/config";

// Note: This Vitest config works for BOTH per-project `test` target AND the
// workspace-wide `coverage` (root-only) target. This configuration worked
// best while I try to figure out best Nx + Vitest integration. See more:
// <https://github.com/vitest-dev/vitest/discussions/4544>.

export default defineConfig({
	plugins: [tsconfigPaths({ projects: [`tsconfig.json`] })],
	test: {
		/* eslint-disable sort-keys -- Prefer intentional grouping */
		clearMocks: true,

		include: [`**/src/**/*.test.ts`],
		exclude: [...configDefaults.exclude, `**/build/**`],
		watchExclude: [...configDefaults.watchExclude, `**/build/**`],

		coverage: {
			clean: true,

			include: [`**/src/**`],
			exclude: [
				...coverageConfigDefaults.exclude,
				`**/{index,exports}.ts`, // Index/exports files should just export
			],

			// These reporters are specific to code coverage
			reporter: [`text`, [`html`, { subdir: `html` }]],
			reportsDirectory: `coverage`,

			provider: `v8`,
			thresholds: {
				branches: 85,
				functions: 100,
				lines: 85,
			},
		},
		/* eslint-enable sort-keys -- END */
	},
});
```

**Reply by connorjs** [reply]
To end my evening exploration, I ended up just going the Vitest happy path route with workspace support. Nx just won’t be involved in test/coverage (yet). I have configured a `root:coverage` task so I can keep using `nx` not the CLI.

Thus, I did end up using `vitest.projects.ts` (renamed from `vitest.workspace.ts` to align with Nx projects term + because I went inline project config route).

```ts
import { basename } from "node:path";

import tsconfigPaths from "vite-tsconfig-paths";
import type { UserWorkspaceConfig } from "vitest/config";
import { defineWorkspace } from "vitest/config";

const projects = [`apps/vite-spa`, `libs/hello-js`];

// eslint-disable-next-line import/no-default-export -- Vitest config
export default defineWorkspace(
	projects.map((project) => vitestProjectConfig(project)),
);

function vitestProjectConfig(project: string): UserWorkspaceConfig {
	return {
		plugins: [tsconfigPaths({ projects: [`tsconfig.json`] })],
		test: {
			environmentMatchGlobs: [[`**/*.tsx`, `happy-dom`]],
			include: [`${project}/src/**/*.test.{ts,tsx}`],
			name: basename(project),
		},
	};
}
```

and `vitest.config.ts` for completeness

```ts
import {
	configDefaults,
	coverageConfigDefaults,
	defineConfig,
} from "vitest/config";

// Note: We are using Vitest’s workspace functionality. See more discussion:
// <https://github.com/vitest-dev/vitest/discussions/4544>.

export default defineConfig({
	test: {
		/* eslint-disable sort-keys -- Prefer intentional grouping */
		clearMocks: true,

		exclude: [...configDefaults.exclude, `**/build/**`],
		watchExclude: [...configDefaults.watchExclude, `**/build/**`],

		// These only apply when Vitest run with `--coverage`
		coverage: {
			clean: true,

			include: [`**/src/**`],
			exclude: [
				...coverageConfigDefaults.exclude,
				`**/{index,exports}.ts`, // Index/exports files should just export
			],

			// These reporters are specific to code coverage
			reporter: [`text-summary`, [`html`, { subdir: `html` }]],
			reportsDirectory: `build/coverage`,

			provider: `v8`,
			thresholds: {
				branches: 85,
				functions: 100,
				lines: 85,
			},
		},
		/* eslint-enable sort-keys -- END */
	},
});
```

**Comment by connorjs** [comment, marked as answer]
Tentatively marking #4561 as the solution.

**I have not configured it with Nx yet** but wanted to make sure I added the link.
